### ISSUE TITLE
Compiler-generated size-0 device dim causes SIGFPE in get_device_stride_infos during Spyre-to-CPU copy

### ISSUE THREAD

**Issue by postmath** [torch-spyre/torch-spyre#3604]
### What happened?

A compiler-generated `SpyreTensorLayout` containing a **zero-sized device dimension** reaches `spyre::get_device_stride_infos`, which divides by it and terminates the Python process with `SIGFPE`.

**Root cause — the layout is generated with a size-0 dim.** `_rescale_stl_for_dtype` at `torch_spyre/_inductor/propagate_layouts.py:357` rescales the stick-indexing dim with floor division:

```python
out_device_size[i] = stl.device_size[i] * in_eps // out_eps
```

For an `fp32 -> fp16` conversion the stick depth doubles (`in_eps=32`, `out_eps=64`), so a stick-group dim of size 1 becomes `1 * 32 // 64 == 0`. The zero then survives the round trip back to fp32 (`0 * 64 // 32 == 0`):

```
[1,4,32] --fp32->fp16--> [0,4,64] --fp16->fp32--> [0,4,32]
```

**Crash — the consumer has no guard for size 0.** In `torch_spyre/csrc/spyre_mem.cpp`, `get_device_stride_infos`:

```cpp
if (tile_size == 1) continue;                          // line 263 - skips 1, not 0
TORCH_CHECK(host_size % elements_before == 0, ...);    // line 266 - SIGFPE here
```

With `tile_size = device_size[0] = 0` the size-1 guard does not fire, so line 281 sets `dcsi_sizes[0] = std::min(remaining_elements, 0) = 0`, line 283 does `elements_before *= 0`, and the next loop iteration evaluates `32 % 0` at line 266.

**Why this is worse than a normal failure.** Integer modulo-by-zero raises a *signal*, not an exception. Neither `try/except` nor `unittest.expectedFailure` can catch it, so it kills the entire pytest process. In a full-suite run it aborted at roughly 38% and masked every test after it. Note also that the underlying case is already marked `XFAIL` on the non-LX path, so a mechanism that ought to be an expected failure is instead a suite-wide abort.

No user error is involved — the malformed layout is produced internally during ordinary compilation.

### What did you expect to happen?

Two independent expectations:

1. **`_rescale_stl_for_dtype` should not emit a size-0 dim.** When `device_size[i] * in_eps` is not divisible by `out_eps`, it should reject with a clear compile-time error rather than flooring to zero. `(4, 32)` fp32->fp16 is already known-unsupported — it is listed in `TO_DTYPE_OP_SHAPES_UNALIGNED` (`tests/inductor/test_inductor_ops.py:335`, commented *"exactly 1 fp32 stick"*) and is `XFAIL` on the non-LX path. The desired outcome is the same clean, catchable failure there, not a crash.

2. **`get_device_stride_infos` should never SIGFPE on a malformed layout.** A guard expression that itself kills the process is strictly worse than a failed check. It should `TORCH_CHECK` (e.g. `elements_before > 0`, and handle `tile_size == 0`) and raise a Python-visible `RuntimeError`.

Fixing only (2) turns the crash into a clean error but leaves the compiler emitting a garbage layout. Fixing only (1) fixes this instance but leaves the crash reachable by other routes.

### How can we reproduce it?

Reproduced on `main` at `193d8a1a` in a clean workspace.

**1. Via the test suite**:

```bash
python3 -m pytest tests/inductor/test_inductor_ops_lx_planning.py \
  -k "round_trip_to_dtype and float32_to_float16 and 4x32" -p no:cacheprovider
```

Crashes on `test_round_trip_to_dtype_add_float32_to_float16_4x32_lx_planning_pointwise`.

**2. The layout bug alone** (no hardware needed, runs in under a second):

```python
import torch
from torch_spyre._C import SpyreTensorLayout, ElementArrangement as EA, get_device_dtype
from torch_spyre._inductor.propagate_layouts import _rescale_stl_for_dtype

stl = SpyreTensorLayout([1, 4, 32], [32, 32, 1], get_device_dtype(torch.float32), EA.STANDARD)
out = _rescale_stl_for_dtype(stl, torch.float16, EA.STANDARD)
print("fp32->fp16:", list(out.device_size), list(out.stride_map))
back = _rescale_stl_for_dtype(out, torch.float32, EA.STANDARD)
print("back->fp32:", list(back.device_size), list(back.stride_map))
```

Output — note the leading `0`, which matches exactly the layout observed in the debugger at the crash:

```
fp32->fp16: [0, 4, 64] [64, 32, 1]
back->fp32: [0, 4, 32] [32, 32, 1]
```

### Any environmental details we need to know?

- `main` at `193d8a1a` ("Small reproducer of test that helps catch issues when implementing coarse tiling dim order changes (#3586)")
- torch `2.13.0+cpu`, Python `3.12.13`
- Default `SENCORES` (32); LX planning enabled via the `test_inductor_ops_lx_planning.py` harness
- **Not** introduced by any in-flight branch: `spyre_mem.cpp`, `propagate_layouts.py`, `test_inductor_ops.py` and `test_inductor_ops_lx_planning.py` are all byte-identical to `main` on the branch where this was first seen, and the crash was independently reproduced in a fresh `main` workspace.

### Anything else we need to know?

**Relationship to #3401 — same crash class, different defect.** #3401 reports a `SIGFPE` in the same function, and its request for early validation with a clear exception would cover expectation (2) above *if implemented comprehensively*. But the two crash on **different divisors at different lines**, verified by reproducing both under gdb:

| | #3401 | This issue |
|---|---|---|
| Crash line | `spyre_mem.cpp:271` | `spyre_mem.cpp:266` |
| Faulting expression | `current_elements / tile_stride` | `host_size % elements_before` |
| Zero divisor | `tile_stride == 0` | `elements_before == 0` |
| Why zero | `host_strides[i]/host_stride` floors to 0 | `tile_size == 0` (size-0 device dim) |
| Direction | `host2device=true` (CPU->Spyre) | `host2device=false` (Spyre->CPU) |
| Origin | User passes malformed layout to public API | Compiler generates it internally |

The layout in #3401 contains no zero at all (`device_size=[1,4,128,64,64]`); its zero arises in a *stride* computation. **A narrow fix for #3401 — validating the public constructor, or guarding only `tile_stride` at line 271 — will not fix this issue.** Conversely, #3401 is framed throughout as *"Invalid user-provided layout metadata should not terminate the process"*, which does not describe this bug, since here no user-supplied layout is involved.

**The hazard is already known and only half-guarded.** `propagate_layouts.py:542-548` documents this exact failure mode:

> *2. Plain conversions (e.g. fp8->fp16 after qfp8ch). Here the input device layout can be degenerate — qfp8ch rescales a size-1 num-sticks dim to 0 (1\*64//128), leaving a size-0 dim — and `_rescale_stl_for_dtype` would faithfully propagate that garbage, changing the layout rank and downstream graph partitioning. Rebuild a clean dense layout from the output host size instead, as the general (non-EA) convert path does.*

That dense-rebuild workaround is applied only to the plain-conversion branch. The staggered branch immediately above it calls `_rescale_stl_for_dtype` unguarded:

```python
if fmt in STAGGERED_EAS or input_ea in STAGGERED_EAS:
    return [_rescale_stl_for_dtype(stl, output.dtype, fmt)]   # line 549-550, no guard
```

That unguarded branch is the path taken here.

**Possible scope.** 70 collected tests in `test_inductor_ops_lx_planning.py`'s `round_trip_to_dtype` family use the unaligned shapes from `TO_DTYPE_OP_SHAPES_UNALIGNED`. Only those whose rescaled dim floors to 0 will crash; one (`float32_to_float16_4x32`) is confirmed, and the rest have not been enumerated because the first crash aborts the run.

**Separate, minor observation.** `_copy_canonical_tests` in `test_inductor_ops_lx_planning.py` does not carry over the non-LX `expect_fail` list, so these unaligned LX round-trip tests run unmarked while their non-LX counterparts are `XFAIL`. Worth a look independently — though note that `expectedFailure` would not have helped here anyway, since a signal cannot be caught.

### Relevant log output

```shell
$ python3 -m pytest tests/inductor/test_inductor_ops_lx_planning.py \
    -k "round_trip_to_dtype and float32_to_float16 and 4x32" -p no:cacheprovider

Fatal Python error: Floating-point exception

Current thread 0x00007f383252d300 (most recent call first):
  File ".../torch_spyre/ops/eager.py", line 247 in spyre__copy_from
  File ".../tests/inductor/utils_inductor.py", line 553 in _to_cpu
  File ".../tests/inductor/utils_inductor.py", line 597 in _compile_and_run
  File ".../tests/inductor/utils_inductor.py", line 686 in compare_with_cpu
  File ".../tests/inductor/test_inductor_ops_lx_planning.py", line 186 in compare_with_cpu
  File ".../tests/inductor/test_inductor_ops.py", line 6512 in test_round_trip_to_dtype_cpu

--- gdb, at the faulting instruction ---

Thread 1 "python3" received signal SIGFPE, Arithmetic exception.
0x00007fd4f196ff0e in spyre::get_device_stride_infos (
    sizes=..., spyre_dma_strides=..., storage_offset=0, stl=...,
    host2device=false, cpu_tensor_strides=...)
    at torch_spyre/csrc/spyre_mem.cpp:266
266       TORCH_CHECK(

(gdb) info locals
tile_index      = 2
tile_size       = 32
j               = 0
host_stride     = 1
host_size       = 32
elements_before = 0          <-- the divisor
i               = 1
tile_map        = {{1}, {2, 0}}
device_rank     = 3

(gdb) output stl.device_size
std::vector of length 3, capacity 3 = {0, 4, 32}     <-- zero-sized dim
(gdb) output stl.stride_map
std::vector of length 3, capacity 3 = {32, 32, 1}
```
